### ISSUE TITLE
Manually rethrowing reverts impedes custom error decoding

### ISSUE THREAD

**Issue by frangio** [NomicFoundation/hardhat#2546]
We sometimes use a [`verifyCallResult`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/02fcc75bb7f35376c22def91b0fb9bc7a50b9458/contracts/utils/Address.sol#L201-L205) helper to rethrow the returndata if a low level call results in an error:

```solidity
        (bool success, bytes memory returndata) = target.call(data);
        return verifyCallResult(success, returndata, errorMessage);
```

With compiler optimizations turned on, if the returndata contains a custom error, Hardhat is not able to decode the rethrown custom error, presumably because it can't identify where the error actually originates because rethrowing is not the standard solc subroutine.

Things work fine with optimizations turned off.

**Comment by frangio** [comment]
After further testing I believe this isn't specific to manual rethrowing, even for built-in rethrowing optimizations cause custom error decoding to fail. (Sorry I'm not providing reproduction steps.)

**Comment by alcuadrado** [comment]
@fvictorio I'm assigning this issue to you, as it seems important, and you are more familiar with this part of the code.

**Comment by frangio** [comment]
Here's a reproduction that shows the source of an error to be wrongly identified due to manually rethrowing. I haven't included this in the reproduction but if custom errors were involved I can imagine how decoding would be broken.

https://github.com/frangio/hardhat-rethrow-reproduction

**Comment by frangio** [comment]
Still relevant.

**Comment by fvictorio** [comment]
Fixed in Hardhat v2.10.1.
